src/value.cpp

```diff
@@ -509,6 +493,68 @@ expr Tensor::to1DArrayWithOfs(
         aop::mkZeroElemFromArr(arr)));
 }
 
+MemRef::Layout::Layout(const vector<expr> &dims):
+    inbounds(init()), mapping(init()), precondition(init()) {
+  vector<expr> indVars, inverseMappings;
+  expr inbounds = mkBool(true);
+
+  for (int i = 0; i < dims.size(); i ++) {
+    indVars.push_back(Index("idx" + to_string(i)));
+    inbounds = inbounds && z3::ult(indVars[i], dims[i]);
+  }
+
+  expr idx = Index("1DIdx");
+  vector<expr> inverseIdxs = from1DIdx(idx, dims);
+  for (auto inverse : inverseIdxs)
+    inverseMappings.push_back(smt::lambda(idx, inverse));
+
+  this->indVars = indVars;
+  this->inbounds = smt::lambda(indVars, inbounds);
+  this->mapping = smt::lambda(indVars, to1DIdx(indVars, dims));
+  this->inverseMappings = inverseMappings;
+  this->precondition = mkBool(true);
+}
+
+MemRef::Layout::Layout(const std::vector<smt::expr> &indVars,
+    const smt::expr &layout,
+    const smt::expr &inbounds,
+    bool useUF): indVars(indVars), inbounds(init()), mapping(init()), precondition(init()) {
+  if (useUF) {
+    vector<smt::sort> domains(indVars.size(), Index::sort());
+    auto layoutFn = mkUF(domains, Index::sort(), freshName("layoutFn"));
+    auto layoutFnExpr = fapply(layoutFn, indVars);
+    expr condition = (layoutFnExpr == layout);
+    vector<expr> inverseMappings;
+    for (unsigned i = 0; i < indVars.size(); i ++) {
+      auto inverseName = freshName("inverse" + to_string(i));
+      auto inverseFn = mkUF(Index::sort(), Index::sort(), move(inverseName));
+      auto inverse = smt::lambda(indVars[i], inverseFn(indVars[i]));
+      inverseMappings.push_back(inverse);
+
+      condition = condition && smt::select(inverse, layoutFnExpr) == indVars[i];
+    }
+    this->inbounds = smt::lambda(indVars, inbounds);
+    this->mapping = smt::lambda(indVars, layoutFnExpr);
+    this->inverseMappings = inverseMappings;
+    this->precondition = smt::forall(indVars, smt::implies(inbounds, condition));
+  } else {
+    expr condition = mkBool(true);
+    vector<expr> inverseMappings;
+    for (unsigned i = 0; i < indVars.size(); i ++) {
```

Comment:
Can we merge this duplicated part in if and else?

Reply:
This is intended to make it easier to remove certain encodings later.